### PR TITLE
More sim handle cleanup

### PR DESCRIPTION
**Pull request by ktbarrett**
First PR of many. This adds a typing stub for `cocotb.simulator` using mypy's stubgen. Makes a couple small improvements to `SimHandleBase`:
* removes `get_definition_file` and `get_definition_name` as they are duplicates of `_def_file` and `_def_name` and conflict with the sub-handle namespace
* removes `_fullname` attribute as it was of little actual utility and probably not being used as intended (`fullname` in the GPI means the full path to the object), so the couple uses were updated to use `_path` instead.
* Improves documentation and typing.

### CODE REVIEW

File: src/cocotb/simulator.pyi [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    def get_range(self) -> tuple[int, int]: ...
+    def get_signal_val_binstr(self) -> str: ...
+    def get_signal_val_long(self) -> int: ...
+    def get_signal_val_real(self) -> float: ...
+    def get_signal_val_str(self) -> bytes: ...
+    def get_type(self) -> int: ...
+    def get_type_string(self) -> str: ...
+    def iterate(self, mode: int) -> gpi_iterator_hdl: ...
+    def set_signal_val_binstr(self, action: int, value: str) -> None: ...
+    def set_signal_val_int(self, *args, **kwargs): ...
```
Comment (marlonjames) [inline src/cocotb/simulator.pyi:56]:
```suggestion
    def set_signal_val_int(self, action: int, value: int): ...
```
Looks like the typed definition is missing from the `PyDoc_STR()` for this method in `simulatormodule.cpp`.

Comment (ktbarrett) [reply src/cocotb/simulator.pyi:56]: I'll fix simulatormodule.cpp as well.

File: src/cocotb/handle.py [hunk 2]
```diff
@@ -64,18 +64,16 @@ def _value_limits(n_bits, limits):
...
     @abstractmethod
-    def __init__(self, handle, path):
-        """
-        .. Constructor. This RST comment works around sphinx-doc/sphinx#6885
-
-        Args:
-            handle (int): The GPI handle to the simulator object.
-            path (str): Path to this handle, ``None`` if root.
-        """
+    def __init__(self, handle: simulator.gpi_sim_hdl, path: str) -> None:
```
Comment (marlonjames) [inline src/cocotb/handle.py:76]: `SimHandle` has a `None` default for **path**. We could do this change or make root an empty string instead.
```suggestion
    def __init__(self, handle: simulator.gpi_sim_hdl, path: Optional[str]) -> None:
```

**Comment by marlonjames** [comment]
We've been a little inconsistent on newsfragments and `versionchanged`/`versionadded`. I think this needs a removal newsfragment for the public `get_definition_*()` methods.

**Comment by ktbarrett** [comment]
Newsfrag and versionchanged note added for `get_defintion_*` methods.

File: docs/source/newsfragments/3609.removal.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+Methods ``get_definition_name()`` and ``get_definition_file()`` of :class:`cocotb.handle.SimHandleBase` were removed in favor of :meth:`~cocotb.handle.SimHandleBase._def_file` and :meth:`~cocotb.handle.SimHandleBase._def_name`, respectively.
```
Comment (marlonjames) [inline docs/source/newsfragments/3609.removal.rst:1]:
```suggestion
Methods ``get_definition_name()`` and ``get_definition_file()`` of :class:`cocotb.handle.SimHandleBase` were removed in favor of :meth:`~cocotb.handle.SimHandleBase._def_name` and :meth:`~cocotb.handle.SimHandleBase._def_file`, respectively.
```

File: src/cocotb/handle.py [hunk 4]
```diff
@@ -64,18 +64,19 @@ def _value_limits(n_bits, limits):
...
-    We maintain a handle which we can use for GPI calls.
+    All simulation objects are hashable and equatable by identity.
+
+    .. code-block:: python3
+        a = dut.clk
+        b = dut.clk
+        assert a == b
+
+    .. versionchanged:: 2.0
+        ``get_definition_name()`` and ``get_definition_file()`` were removed in favor of :meth:`_def_file` and :meth:`_def_name`, respectively.
```
Comment (marlonjames) [inline src/cocotb/handle.py:75]:
```suggestion
        ``get_definition_name()`` and ``get_definition_file()`` were removed in favor of :meth:`_def_name` and :meth:`_def_file`, respectively.
```

**Comment by ktbarrett** [comment]
I sure am firing on all cylinders this day... week... month... life.

**Review by marlonjames — APPROVED** [review]

**Pull request merged**